homeassistant/components/climate/zhong_hong.py

```diff
@@ -0,0 +1,205 @@
+"""
+Support for ZhongHong HVAC Controller.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/climate.zhong_hong/
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.climate import (
+    ATTR_OPERATION_MODE, PLATFORM_SCHEMA, STATE_COOL, STATE_DRY,
+    STATE_FAN_ONLY, STATE_HEAT, SUPPORT_FAN_MODE, SUPPORT_ON_OFF,
+    SUPPORT_OPERATION_MODE, SUPPORT_TARGET_TEMPERATURE, ClimateDevice)
+from homeassistant.const import (
+    ATTR_TEMPERATURE, CONF_HOST, CONF_PORT, EVENT_HOMEASSISTANT_START,
+    EVENT_HOMEASSISTANT_STOP, TEMP_CELSIUS)
+from homeassistant.exceptions import PlatformNotReady
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_GATEWAY_ADDRRESS = 'gateway_address'
+
+REQUIREMENTS = ['zhong_hong_hvac==1.0.9']
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST):
+    cv.string,
+    vol.Optional(CONF_PORT, default=9999):
+    vol.Coerce(int),
+    vol.Optional(CONF_GATEWAY_ADDRRESS, default=1):
+    vol.Coerce(int),
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the ZhongHong HVAC platform."""
+    from zhong_hong_hvac.hub import ZhongHongGateway
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
+    gw_addr = config.get(CONF_GATEWAY_ADDRRESS)
+    hub = ZhongHongGateway(host, port, gw_addr)
+    try:
+        devices = [
+            ZhongHongClimate(hub, addr_out, addr_in)
+            for (addr_out, addr_in) in hub.discovery_ac()
+        ]
+    except Exception as exc:
+        _LOGGER.error("ZhongHong controller is not ready", exc_info=exc)
+        raise PlatformNotReady
+
+    def startup(event):
+        """Add devices to HA and start hub socket."""
+        add_devices(devices)
```

Comment:
Why do we wait with adding the devices until home assistant start?

Reply:
Do you mean I should add the devices first. And start listen when home assistant start?

Reply:
I don't see a reason to wait with adding the devices at least. What's the reason to wait with start listening?

You know the device library, there might be good reasons to wait. I'm just asking if it's necessary. In general it's best to have everything setup and ready before home assistant starts. That will allow users to listen to the start event and take action directly.

Reply:
Main reason to add them during setup is because automations will be activated when Home Assistant starts. They can now be triggered by just devices being added.

Reply:
I think the decision you make is more important than mine.

Reply:
If there is no bad things to happen, I intend to put `add_devices` and `start_listen` right in the `setup_platform` stage.

Reply:
Emm. I found that `start_listen` should be invoked after all zhong_hong climate entity is setted up(after all entity's `async_added_to_hass` is called).

So is this the good time to `start_listen` after `EVENT_HOMEASSISTANT_START` event fired?